vmm/task/src/youki.rs

```diff
@@ -0,0 +1,646 @@
+        Err(Error::Unimplemented("update resource".to_string()))
+    }
+
+    #[cfg(target_os = "linux")]
+    async fn stats(&self, p: &InitProcess) -> Result<Metrics> {
+        if p.pid <= 0 {
+            return Err(other!(
+                "failed to collect metrics because init process is {}",
+                p.pid
+            ));
+        }
+        containerd_shim::cgroup::collect_metrics(p.pid as u32)
+    }
+
+    #[cfg(not(target_os = "linux"))]
+    async fn stats(&self, _p: &InitProcess) -> Result<Metrics> {
+        Err(Error::Unimplemented("process stats".to_string()))
+    }
+
+    async fn ps(&self, p: &InitProcess) -> Result<Vec<ProcessInfo>> {
+        Ok(vec![ProcessInfo {
+            pid: p.pid as u32,
+            ..Default::default()
+        }])
+    }
+}
+
+impl YoukiInitLifecycle {
+    pub fn new(youki_container: Arc<Mutex<Container>>, opts: Options, bundle: &str) -> Self {
+        let work_dir = Path::new(bundle).join("work");
+        let mut opts = opts;
+        if opts.criu_path().is_empty() {
+            opts.criu_path = work_dir.to_string_lossy().to_string();
+        }
+        Self {
+            youki_container,
+            exit_signal: Default::default(),
+        }
+    }
+}
+
+pub struct YoukiExecLifecycle {
+    bundle: String,
+    container_id: String,
+    io_uid: u32,
+    io_gid: u32,
+    spec: Process,
+    exit_signal: Arc<ExitSignal>,
+}
+
+#[async_trait]
+impl ProcessLifecycle<ExecProcess> for YoukiExecLifecycle {
+    async fn start(&self, p: &mut ExecProcess) -> containerd_shim::Result<()> {
+        rescan_pci_bus().await?;
+        let (socket, pio, _container_io) = if p.stdio.terminal {
+            let s = ConsoleSocket::new().await?;
+            (Some(s), None, None)
+        } else {
+            let (pio, container_io) = create_io(&p.id, self.io_uid, self.io_gid, &p.stdio)?;
+            //exec_opts.io = pio.io.as_ref().cloned();
+            (None, Some(pio), Some(container_io))
+        };
+
+        let probe_path = format!("{}/{}-process.json", self.bundle, &p.id);
+        let spec_str = serde_json::to_string(&self.spec)
+            .map_err(other_error!(e, "failed to marshall exec spec to string"))?;
+        tokio::fs::write(&probe_path, &spec_str)
+            .await
+            .map_err(other_error!(e, "failed to write spec to process.json"))?;
+
+        let container_id = self.container_id.clone();
+        let socket_path = socket.as_ref().map(|p| p.path.clone());
+        let probe_path_clone = probe_path.clone();
+        let exec_result = spawn_blocking(move || {
+            let builder = ContainerBuilder::new(container_id, SyscallType::default())
+                .with_root_path(PathBuf::from(YOUKI_DIR))
+                .map_err(other_error!(e, "failed to set youki root path"))?
+                .with_console_socket(socket_path);
+
+            // if let Some(f) = container_io {
```

Comment:
Uncomment this after https://github.com/youki-dev/youki/pull/2961 is merged into youki